client/core/core.go

```diff
@@ -1660,10 +1659,11 @@ func (c *Core) ReconfigureWallet(appPW, newWalletPW []byte, assetID uint32, cfg
 	}
 	c.connMtx.RUnlock()
 
-	if oldWallet.connected() {
-		oldWallet.Disconnect()
-	}
+	c.walletMtx.Lock()
 	c.wallets[assetID] = wallet
+	c.walletMtx.Unlock()
+
+	go oldWallet.Disconnect()
```

Comment:
Even with the deadlock resolved by not holding the walletMtx at this point, this can be done lazily.

Reply:
I see that in **runner.go**, we have

```
func (c *ConnectionMaster) Disconnect() {
	c.cancel()
	c.mtx.RLock()
	c.wg.Wait()
	c.mtx.RUnlock()
}
```

Should that be 

```
func (c *ConnectionMaster) Disconnect() {
	c.cancel()
	c.mtx.RLock()
	wg := c.wg
	c.mtx.RUnlock()
	wg.Wait()
}
```

so that we wouldn't have to do this?

Reply:
It would still be stuck at `wg.Wait()`.  The ConnectionMaster's mutex isn't under contention.